### PR TITLE
[dev] performance on GPU 

### PR DESCRIPTION
**Pull request by cnpetra**
This PR 

1. ensures robust performance for the GPU linear solves by switching between fast/no pivoting  and  stable linear (GPU) solvers at runtime when numerical issues are encountered (loss of accuracy in the linear solves or search directions of negative curvature)
 
2. adds support for additional built-in performance profiling for the solution of various KKT  linear systems in HiOp

Feature 2 can be turned on by using the newly added `time_kkt` (string) option either programatically or by adding the line  
``` 
time_kkt on
```
to  `hiop.options` file.

Tested/Developed on Summit. Passes all tests both on CPU and GPU.

### CODE REVIEW

**Comment by pelesh** [comment]
When I try to build on Mac I get following linker error:
```bash
[ 56%] Linking CXX executable test_hiopLinAlgComplex.exe
Undefined symbols for architecture x86_64:
  "typeinfo for hiop::hiopOptions", referenced from:
      l___unnamed_15 in hiopLinSolver.cpp.o
  "typeinfo for hiop::hiopNlpFormulation", referenced from:
      l___unnamed_12 in hiopLinSolver.cpp.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [src/LinAlg/test_hiopLinAlgComplex.exe] Error 1
make[1]: *** [src/LinAlg/CMakeFiles/test_hiopLinAlgComplex.exe.dir/all] Error 2
make: *** [all] Error 2
```
I am building with clang++ 9.0

**Comment by pelesh** [comment]
The code builds cleanly on Mac with GCC 9.3.0 and passes all tests.

**Comment by pelesh** [comment]
It seems there is a CMake bug related to configuration of linear solver interface build. I get following compilation error:
```bash
[ 31%] Building CXX object src/LinAlg/CMakeFiles/hiopLinAlgGPU.dir/hiopLinSolverIndefDenseMagma.cpp.o
cd /.../src/hiop/build/src/LinAlg && /share/apps/gcc/7.4.0/bin/g++   -I/.../rh7p9/openmpi/3.1.5/gcc/7.4.0/include -I/.../src/hiop/build -I/.../src/hiop/hiop/src/Interface -I/.../src/hiop/hiop/src/Optimization -I/.../src/hiop/hiop/src/LinAlg -I/.../src/hiop/hiop/src/Utils -I/.../src/hiop/hiop/src/_Internals  -g -fPIC   -std=gnu++11 -o CMakeFiles/hiopLinAlgGPU.dir/hiopLinSolverIndefDenseMagma.cpp.o -c /.../src/hiop/hiop/src/LinAlg/hiopLinSolverIndefDenseMagma.cpp
In file included from /.../src/hiop/hiop/src/LinAlg/hiopLinSolverIndefDenseMagma.cpp:1:0:
/.../src/hiop/hiop/src/LinAlg/hiopLinSolverIndefDenseMagma.hpp:9:10: fatal error: magma_v2.h: No such file or directory
 #include "magma_v2.h"
          ^~~~~~~~~~~~
compilation terminated.
```
Include flag for Magma is not set.

File: src/LinAlg/CMakeLists.txt [hunk 1]
```diff
@@ -2,13 +2,19 @@ add_library(hiopLinAlg
...
         hiopMatrixComplexDense.cpp
         hiopMatrixSparseTripletStorage.cpp
         hiopMatrixSparseTriplet.cpp
         hiopMatrixComplexSparseTriplet.cpp)
 target_link_libraries(hiopLinAlg PUBLIC hiopOptimization hiop_math)
 
+if(HIOP_USE_MAGMA)
+  add_library(hiopLinAlgGPU OBJECT hiopLinSolverIndefDenseMagma.cpp)
+  target_sources(hiopLinAlg PUBLIC $<TARGET_OBJECTS:hiopLinAlgGPU>)
+endif(HIOP_USE_MAGMA)
```
[the remark this answers is not in the document]
Comment (pelesh) [reply src/LinAlg/CMakeLists.txt:16]: I think you also need to explicitly link HiOp with `hiopLinAlgGPU` in the top level CMake file.

**Comment by pelesh** [comment]
With PR #69 merged, this PR is ready to be merged, as well.

**Comment by cnpetra** [comment]
> When I try to build on Mac I get following linker error:
> 
> ```shell
> [ 56%] Linking CXX executable test_hiopLinAlgComplex.exe
> Undefined symbols for architecture x86_64:
>   "typeinfo for hiop::hiopOptions", referenced from:
>       l___unnamed_15 in hiopLinSolver.cpp.o
>   "typeinfo for hiop::hiopNlpFormulation", referenced from:
>       l___unnamed_12 in hiopLinSolver.cpp.o
> ld: symbol(s) not found for architecture x86_64
> clang: error: linker command failed with exit code 1 (use -v to see invocation)
> make[2]: *** [src/LinAlg/test_hiopLinAlgComplex.exe] Error 1
> make[1]: *** [src/LinAlg/CMakeFiles/test_hiopLinAlgComplex.exe.dir/all] Error 2
> make: *** [all] Error 2
> ```
> 
> I am building with clang++ 9.0

I got something similar on a mac system. I think the fix is as in https://github.com/LLNL/hiop/pull/70

**Pull request merged**